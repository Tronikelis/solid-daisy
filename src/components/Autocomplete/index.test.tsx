/* eslint-disable solid/reactivity */
import { render } from "@solidjs/testing-library";
import userEvent from "@testing-library/user-event";
import { ComponentProps, createSignal } from "solid-js";
import { beforeEach, describe, expect, it, vi } from "vitest";

import { Autocomplete } from ".";

describe("<Autocomplete />", () => {
    let props: ComponentProps<typeof Autocomplete>;

    beforeEach(() => {
        props = {
            items: [],
            value: [],
            setValue: vi.fn(),
        };
    });

    it("renders correctly", () => {
        const { container } = render(() => <Autocomplete {...props} />);
        expect(container).toMatchSnapshot();
    });

    it("dropdown opens on input focus and closes on blur", async () => {
        const screen = render(() => (
            <>
                <div>outside</div>
                <Autocomplete {...props} data-testid="input" />
            </>
        ));

        await userEvent.click(screen.getByTestId("input"));

        expect(screen.getByTestId("dropdown")).toHaveClass("visible");

        await userEvent.click(screen.getByText("outside"));

        expect(screen.getByTestId("dropdown")).toHaveClass("invisible");
    });

    describe("with items", () => {
        beforeEach(() => {
            props.items = [{ value: "foobar" }, { value: "let's go" }];
        });

        it("dropdown shows items", () => {
            const screen = render(() => <Autocomplete {...props} />);

            expect(screen.getByText("foobar")).toBeInTheDocument();
            expect(screen.getByText("let's go")).toBeInTheDocument();
        });

        it("selecting item works", async () => {
            const [value, setValue] = createSignal<{ value: string }[]>([]);

            const screen = render(() => (
                <Autocomplete {...props} value={value()} setValue={setValue} />
            ));

            await userEvent.click(screen.getByText("foobar"));

            expect(screen.getAllByText("foobar")).toHaveLength(2);
            expect(value()).toStrictEqual([{ value: "foobar" }]);

            await userEvent.click(screen.getAllByText("foobar")[1]!);

            expect(value()).toStrictEqual([]);
            expect(screen.getAllByText("foobar")).toHaveLength(1);
        });

        it("default filtering works", async () => {
            props.items = [{ value: "foobar" }, { value: "another" }];

            const screen = render(() => <Autocomplete {...props} data-testid="input" />);

            await userEvent.type(screen.getByTestId("input"), "foobar");

            expect(screen.getAllByText("foobar")).toHaveLength(1);
            expect(screen.queryByText("another")).not.toBeInTheDocument();
        });

        describe("multiple", () => {
            beforeEach(() => {
                props.multiple = true;
            });

            it("x icon appears to delete item", async () => {
                const [value, setValue] = createSignal([{ value: "foobar" }]);

                const screen = render(() => (
                    <Autocomplete {...props} value={value()} setValue={setValue} />
                ));

                expect(value()).toHaveLength(1);
                expect(screen.getAllByText("foobar")).toHaveLength(2);

                await userEvent.click(screen.getByText("X"));

                expect(value()).toHaveLength(0);

                expect(screen.getAllByText("foobar")).toHaveLength(1);
            });

            it("custom onDelete works", async () => {
                const [value, setValue] = createSignal([{ value: "foobar" }]);

                const screen = render(() => (
                    <Autocomplete
                        {...props}
                        value={value()}
                        setValue={setValue}
                        selectedComponent={(_, onDelete) => (
                            <div onClick={onDelete} data-testid="item">
                                X
                            </div>
                        )}
                    />
                ));

                await userEvent.click(screen.getByTestId("item"));

                expect(value()).toHaveLength(0);
            });
        });

        it("can navigate with arrow keys", async () => {
            const [value, setValue] = createSignal<{ value: string }[]>([]);

            const screen = render(() => (
                <Autocomplete
                    {...props}
                    value={value()}
                    setValue={setValue}
                    data-testid="input"
                />
            ));

            await userEvent.keyboard("[ArrowDown]");

            expect(screen.getByText("foobar").parentElement).not.toHaveClass("ring");

            await userEvent.click(screen.getByTestId("input"));
            await userEvent.keyboard("[ArrowDown]");

            expect(screen.getByText("foobar").parentElement).toHaveClass("ring");

            await userEvent.keyboard("[Enter]");

            expect(value()).toStrictEqual([{ value: "foobar" }]);

            await userEvent.keyboard("[Escape]");

            expect(screen.getByTestId("dropdown")).toHaveClass("invisible");
        });
    });
});
