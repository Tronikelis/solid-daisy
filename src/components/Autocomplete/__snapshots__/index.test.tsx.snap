// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Autocomplete /> > renders correctly 1`] = `
<div>
  <div
    class="flex flex-col gap-0"
  >
    <div
      class="flex flex-col gap-4 input h-auto input-bordered min-h-12 py-2 justify-center text-inherit"
    >
      <div
        class="flex flex-row items-center cursor-[inherit] flex-wrap gap-1"
      >
        <input
          class="pointer-events-auto min-w-6 flex-1 cursor-[inherit] border-none bg-transparent outline-none placeholder-shown:text-ellipsis"
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    class="fixed left-0 top-0 z-[3000] rounded-box border-2 border-base-200 bg-base-100 shadow transition-opacity duration-200 invisible opacity-0"
    data-testid="dropdown"
  >
    <ul
      class="menu max-h-48 flex-nowrap gap-1 overflow-auto"
    />
  </div>
</div>
`;
